api/auth.go

```diff
@@ -0,0 +1,105 @@
+package api
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/SkynetLabs/skynet-accounts/database"
+	"github.com/SkynetLabs/skynet-accounts/jwt"
+	jwt2 "github.com/lestrrat-go/jwx/jwt"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+// userAndTokenByRequestToken scans the request for an authentication token,
+// fetches the corresponding user from the database and returns both user and
+// token.
+func (api *API) userAndTokenByRequestToken(req *http.Request) (*database.User, jwt2.Token, error) {
+	token, err := tokenFromRequest(req)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "error fetching token from request")
+	}
+	sub, _, _, err := jwt.TokenFields(token)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "error decoding token from request")
+	}
+	u, err := api.staticDB.UserBySub(req.Context(), sub)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "error fetching user from database")
+	}
+	return u, token, nil
+}
+
+// userAndTokenByAPIKey extracts the APIKey from the request and validates it.
+// It then returns the user who owns it and a token for that user.
+// It first checks the headers and then the query.
+// This method accesses the database.
+func (api *API) userAndTokenByAPIKey(req *http.Request) (*database.User, jwt2.Token, error) {
+	ak, err := apiKeyFromRequest(req)
+	if err != nil {
+		return nil, nil, err
+	}
+	akr, err := api.staticDB.APIKeyByKey(req.Context(), ak.String())
+	if err != nil {
+		return nil, nil, err
+	}
+	// If we're dealing with a public API key, we need to validate that this
+	// request is a GET for a covered skylink.
+	if akr.Public {
```

Comment:
Why would there be? This is called from `withAuth` which is a middleware we use with all types of requests.

Reply:
yeah just wanted to make sure there's no `PUT`, `POST`, `PATCH` or `DELETE` routes that have a skylink hash in the request URI, but you can resolve if it's all fine, just double-checking because you explicitly mention a `GET` in the comment and then not check `req.Method`

Reply:
Oh, I see. Yes, we can check that here and exit early when the request is not a GET, as that is guaranteed to fail later.
I'll do it in a F/U. 👍

Reply:
Actually, I just did it here because I had to make some changes anyway.